src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs

```diff
@@ -5971,33 +5970,32 @@ .locals init (bool V_0)
       IL_001a:  ldc.i4.0
       IL_001b:  stloc.0
       IL_001c:  ldloc.0
-      IL_001d:  brtrue.s   IL_0021
-      IL_001f:  ldc.i4.0
-      IL_0020:  ret
-      IL_0021:  ldc.i4.1
-      IL_0022:  ret
+      IL_001d:  ret
     }
 ");
             compVerifier.VerifyIL("C.M2", @"
     {
-      // Code size       24 (0x18)
+      // Code size       28 (0x1c)
       .maxstack  2
       IL_0000:  ldarg.0
       IL_0001:  ldc.i4.s   65
       IL_0003:  blt.s      IL_000a
       IL_0005:  ldarg.0
       IL_0006:  ldc.i4.s   90
-      IL_0008:  ble.s      IL_0016
+      IL_0008:  ble.s      IL_001a
       IL_000a:  ldarg.0
       IL_000b:  ldc.i4.s   97
-      IL_000d:  blt.s      IL_0014
+      IL_000d:  blt.s      IL_0018
       IL_000f:  ldarg.0
       IL_0010:  ldc.i4.s   122
-      IL_0012:  ble.s      IL_0016
+      IL_0012:  cgt
       IL_0014:  ldc.i4.0
-      IL_0015:  ret
-      IL_0016:  ldc.i4.1
+      IL_0015:  ceq
       IL_0017:  ret
+      IL_0018:  ldc.i4.0
+      IL_0019:  ret
+      IL_001a:  ldc.i4.1
+      IL_001b:  ret
     }
 ");
         }
```

Comment:
`Correctness_Rebuild` fails with weird errors ([AzDo build](https://dev.azure.com/dnceng-public/public/_build/results?buildId=194570)).

1. It crashes with OOM (deterministically, can even reproduce that locally).

   <details>
   <summary>Details:</summary>

   ```log
   2023-03-07T14:25:15.5083562Z Validation failed for D:\a\_work\1\s\artifacts/obj/Microsoft.CodeAnalysis.ExpressionCompiler\Release\netstandard2.0\Microsoft.CodeAnalysis.ExpressionEvaluator.ExpressionCompiler.dll
   2023-03-07T14:25:15.5084335Z Writing diffs to "D:\a\_work\1\s\artifacts\BuildValidator\netstandard2.0\Microsoft.CodeAnalysis.ExpressionEvaluator.ExpressionCompiler"
   2023-03-07T14:25:16.6559691Z Exception of type 'System.OutOfMemoryException' was thrown.
   2023-03-07T14:25:17.2868546Z Unhandled exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.BadImageFormatException: Exception of type 'System.OutOfMemoryException' was thrown.
   2023-03-07T14:25:17.2912239Z    at System.Reflection.PortableExecutable.PEReader.DecodeEmbeddedPortablePdbDebugDirectoryData(AbstractMemoryBlock block)
   2023-03-07T14:25:17.2913380Z    at System.Reflection.PortableExecutable.PEReader.ReadEmbeddedPortablePdbDebugDirectoryData(DebugDirectoryEntry entry)
   2023-03-07T14:25:17.2914522Z    at System.Reflection.PortableExecutable.PEReader.TryOpenEmbeddedPortablePdb(DebugDirectoryEntry embeddedPdbEntry, Boolean& openedEmbeddedPdb, MetadataReaderProvider& provider, Exception& errorToReport)
   2023-03-07T14:25:17.2915533Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2916473Z    at System.Reflection.PortableExecutable.PEReader.TryOpenAssociatedPortablePdb(String peImagePath, Func`2 pdbFileStreamProvider, MetadataReaderProvider& pdbReaderProvider, String& pdbPath)
   2023-03-07T14:25:17.2917705Z    at BuildValidator.Program.ValidateFile(Options options, AssemblyInfo assemblyInfo, ILogger logger, LocalReferenceResolver referenceResolver) in /_/src/Tools/BuildValidator/Program.cs:line 278
   2023-03-07T14:25:17.2918862Z    at BuildValidator.Program.ValidateFiles(IEnumerable`1 assemblyInfos, Options options, ILoggerFactory loggerFactory) in /_/src/Tools/BuildValidator/Program.cs:line 203
   2023-03-07T14:25:17.2920092Z    at BuildValidator.Program.HandleCommand(String[] assembliesPath, String[] exclude, String sourcePath, String[] referencesPath, Boolean verbose, Boolean quiet, Boolean debug, String debugPath) in /_/src/Tools/BuildValidator/Program.cs:line 140
   2023-03-07T14:25:17.2921119Z    --- End of inner exception stack trace ---
   2023-03-07T14:25:17.2921791Z    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   2023-03-07T14:25:17.2922646Z    at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   2023-03-07T14:25:17.2923305Z    at System.Delegate.DynamicInvokeImpl(Object[] args)
   2023-03-07T14:25:17.2923909Z    at System.CommandLine.Invocation.ModelBindingCommandHandler.<InvokeAsync>d__12.MoveNext()
   2023-03-07T14:25:17.2924532Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2925172Z    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   2023-03-07T14:25:17.2925877Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   2023-03-07T14:25:17.2926703Z    at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
   2023-03-07T14:25:17.2927388Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2928302Z    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   2023-03-07T14:25:17.2928941Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   2023-03-07T14:25:17.2929653Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseErrorReporting>b__21_0>d.MoveNext()
   2023-03-07T14:25:17.2930293Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2930858Z    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   2023-03-07T14:25:17.2931547Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   2023-03-07T14:25:17.2932274Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
   2023-03-07T14:25:17.2933007Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2933583Z    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   2023-03-07T14:25:17.2934255Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   2023-03-07T14:25:17.2935036Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseVersionOption>b__0>d.MoveNext()
   2023-03-07T14:25:17.2935862Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2936465Z    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   2023-03-07T14:25:17.2937091Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   2023-03-07T14:25:17.2937850Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseTypoCorrections>b__0>d.MoveNext()
   2023-03-07T14:25:17.2938507Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2939096Z    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   2023-03-07T14:25:17.2939726Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   2023-03-07T14:25:17.2940461Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__22_0>d.MoveNext()
   2023-03-07T14:25:17.2941157Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2941704Z    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   2023-03-07T14:25:17.2942359Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   2023-03-07T14:25:17.2943033Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseDirective>b__20_0>d.MoveNext()
   2023-03-07T14:25:17.2943675Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2944209Z    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   2023-03-07T14:25:17.2944873Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   2023-03-07T14:25:17.2945555Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseDebugDirective>b__11_0>d.MoveNext()
   2023-03-07T14:25:17.2946187Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2946767Z    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   2023-03-07T14:25:17.2947395Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   2023-03-07T14:25:17.2948145Z    at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
   2023-03-07T14:25:17.2948773Z --- End of stack trace from previous location where exception was thrown ---
   2023-03-07T14:25:17.2949358Z    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   2023-03-07T14:25:17.2949983Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   ```

   </details>

2. Before it crashes, `BuildValidator.exe` also reports different IL after rebuild in test classes - seems exactly the test classes that this PR touches. Constants are emitted differently, I have no idea why.

   <details>
   <summary>Details:</summary>

   For example, in `PatternSwitchTests.DuplicateDouble`:

   ```diff
   -    IL_000a:  /* 23   | 40091EB851EB851F */ ldc.r8     3.1400000000000001
   +    IL_000a:  /* 23   | 40091EB951EB851F */ ldc.r8     3.1400019073486329
   ```

   In `AttributeTests.TestAttributesOnPropertyAndGetSet`:

   ```diff
   @@ -2331,7 +2331,7 @@
          IL_00fc:  /* A2   |                  */ stelem.ref
          IL_00fd:  /* 25   |                  */ dup
          IL_00fe:  /* 19   |                  */ ldc.i4.3
   -      IL_00ff:  /* 23   | 400921FB4D12D84A */ ldc.r8     3.1415926000000001
   +      IL_00ff:  /* 23   | 400921FC4D12D84A */ ldc.r8     3.1415945073486329
          IL_0108:  /* 8C   | (01)0002D0       */ box        [mscorlib/*23000001*/]System.Double/*010002D0*/
        + "}\r\n1.0m !\r\nnull" /* 7036B889 */
   -    IL_0036:  /* 23   | 4000CCCCCCCCCCCD */ ldc.r8     2.1000000000000001
   +    IL_0036:  /* 23   | 4000CCCDCCCCCCCD */ ldc.r8     2.1000019073486329
        IL_003f:  /* 8C   | (01)0002D0       */ box        [mscorlib/*23000001*/]System.Double/*010002D0*/
        IL_0044:  /* 22   | 40066666         */ ldc.r4     2.0999999
        IL_0049:  /* 8C   | (01)0002D1       */ box        [mscorlib/*23000001*/]System.Single/*010002D1*/
   @@ -595932,7 +595932,7 @@
        IL_0050:  /* 8D   | (01)000014       */ newarr     [mscorlib/*23000001*/]System.Object/*01000014*/
        IL_0055:  /* 25   |                  */ dup     IL_0056:  /* 16   |                  */ ldc.i4.0
   -    IL_0057:  /* 23   | 3FF028F5C28F5C29 */ ldc.r8     1.01
   +    IL_0057:  /* 23   | 3FF028F6C28F5C29 */ ldc.r8     1.0100009536743164
        IL_0060:  /* 0D   |                  */ stloc.3     IL_0061:  /* 12   | 03               */ ldloca.s   V_3
        IL_0063:  /* 28   | (0A)0005CA       */ call       instance string [mscorlib/*23000001*/]System.Double/*010002D0*/::ToString() /* 0A0005CA */
   ```

   </details>

Reply:
Okay I figured out the OOM issue. The OOM is coming from a call to `AllocHGlobal` deep inside the `PEStream` loading code. The problem with `AllocHGlobal` is that it's a native memory allocation function and hence doesn't represent any type of pressure to the GC. While we have memory available here, it's just not collected, there isn't enough to satisfy the 41MB request that happens because allocating native memory doesn't pressure the GC to do a collect. 

I _believe_ this is happening in your case because the earlier failures in the execution cause more memory to get loaded (to dump XML, MDV, etc ...). That pushes us over the limit and causes a crash. 

Short term I fixed this by just making the app run as 64 bit locally. Still looking at the actual errors.

Reply:
Pushed a test change to your PR and yes that fixed the issue. Will open a separate PR for that but will keep it here so it's not blocking you.